libs/langchain/langchain/document_loaders/sharepoint.py

```diff
@@ -0,0 +1,59 @@
+"""Loader that loads data from Sharepoint Document Library"""
+from __future__ import annotations
+
+from typing import Iterator, List, Optional, Sequence
+
+from langchain.docstore.document import Document
+from langchain.document_loaders.base_o365 import (
+    O365BaseLoader,
+    _FileType,
+)
+from langchain.document_loaders.parsers.registry import get_parser
+from langchain.pydantic_v1 import Field
+
+
+class SharePointLoader(O365BaseLoader):
+    """Load  from `SharePoint`."""
+
+    document_library_id: str = Field(...)
+    """ The ID of the SharePoint document library to load data from."""
+    folder_path: Optional[str] = None
+    """ The path to the folder to load data from."""
+    object_ids: Optional[List[str]] = None
+    """ The IDs of the objects to load data from."""
+
+    @property
+    def _file_types(self) -> Sequence[_FileType]:
+        """Return supported file types."""
+        return _FileType.DOC, _FileType.DOCX, _FileType.PDF
+
+    @property
+    def _scopes(self) -> List[str]:
+        """Return required scopes."""
+        return ["sharepoint", "basic"]
+
+    def lazy_load(self) -> Iterator[Document]:
+        """Load documents lazily. Use this when working at a large scale."""
+        try:
+            from O365.drive import Drive, Folder
+        except ImportError:
+            raise ImportError(
+                "O365 package not found, please install it with `pip install o365`"
+            )
+        drive = self._auth().storage().get_drive(self.document_library_id)
```

Comment:
could move data fetching to base o365, and replace the share point loader with the generic loader which will allow users to swap out parsing strategies